grudge/tools.py

```diff
@@ -235,4 +239,105 @@ def build_jacobian(actx, f, base_state, stepsize):
+    shape ``(2, 2)`` and an input array *ary* of shape ``(3, 2)``, the call::
+
+        map_subarrays(f, (2,), (2, 2), ary)
+
+    will produce an array of shape ``(3, 2, 2)`` containing the results of calling
+    *f* on the 3 subarrays of shape ``(2,)`` in *ary*. The call::
+
+        map_subarrays(f, (2,), (2, 2), ary, return_nested=True)
+
+    will instead produce an array of shape ``(3,)`` with each entry containing an
+    array of shape ``(2, 2)``.
+
+    :param f: the function to be called.
+    :param in_shape: the shape of any inputs to *f*.
+    :param out_shape: the shape of the result of calling *f* on an array of shape
+        *in_shape*.
+    :param ary: a :class:`numpy.ndarray` instance.
+    :param return_nested: if *out_shape* is nontrivial, this flag indicates whether
+        to return a nested array (containing one entry for each application of *f*),
+        or to return a single, higher-dimensional array.
+
+    :returns: an array with the subarrays of shape *in_shape* replaced with
+        subarrays of shape *out_shape* resulting from the application of *f*.
+    """
+    if not isinstance(ary, np.ndarray):
+        if len(in_shape) != 0:
+            raise ValueError(f"found scalar, expected array of shape {in_shape}")
+        return f(ary)
+    else:
+        if (
+                ary.ndim < len(in_shape)
+                or ary.shape[ary.ndim-len(in_shape):] != in_shape):
+            raise ValueError(
+                f"array of shape {ary.shape} is incompatible with function "
+                f"expecting input shape {in_shape}")
+        base_shape = ary.shape[:ary.ndim-len(in_shape)]
+        if len(base_shape) == 0:
+            return f(ary)
+        else:
+            in_slice = tuple(slice(0, n) for n in in_shape)
+            out_slice = tuple(slice(0, n) for n in out_shape)
+            if return_nested:
+                result = np.empty(base_shape, dtype=ary.dtype)
+                for idx in np.ndindex(base_shape):
+                    result[idx] = f(ary[idx + in_slice])
+                return result
+            else:
+                result = np.empty(base_shape + out_shape, dtype=ary.dtype)
+                for idx in np.ndindex(base_shape):
+                    result[idx + out_slice] = f(ary[idx + in_slice])
+                return result
+
+
+def rec_map_subarrays(
+        f: Callable[[np.ndarray], np.ndarray],
+        in_shape: Tuple[int, ...],
+        out_shape: Tuple[int, ...],
+        ary: np.ndarray, *,
+        leaf_cls: Optional[type] = None,
+        return_nested: bool = False) -> np.ndarray:
+    r"""
+    Like :func:`map_subarrays`, but with support for
+    :class:`arraycontext.ArrayContainer`\ s.
+
+    :param leaf_cls: An array container of this type will be considered a leaf
+        and will be passed to *f* without further destructuring.
+    """
+    if type(ary) is leaf_cls or isinstance(ary, np.ndarray):
```

Comment:
I don't think the `isinstance(ary, np.ndarray)` here is specific enough. It should only call `map_subarrays` if `ary` is an array of scalars. For example, say you have the gradient of a vector of `DOFArray`s that was created using `return_nested=True`; it should look like:

```python
ndarray([
    ndarray([DOFArray, DOFArray, DOFArray]),
    ndarray([DOFArray, DOFArray, DOFArray]),
    ndarray([DOFArray, DOFArray, DOFArray])])
```

If you then take the divergence of that, `rec_map_subarrays` should recurse once on the outer `ndarray` and then pass each of the subarrays to `map_subarrays`. But with this check I think it will pass the whole thing to `map_subarrays`.

Reply:
Fixed it.